### ISSUE TITLE
Babel transform plugin not working within NuxtJS

### ISSUE THREAD

**Issue by babakness** [nuxt/nuxt.js#1776]
Hello, I'm trying to use the newly approved to Stage-1 pipeline operator in a project.

https://github.com/SuperPaintman/babel-plugin-transform-pipeline

https://github.com/babel/proposals/issues/29

I have the following 

```javascript
babel: {  
      "presets": [
        "es2015",
        "stage-2"
      ],
      "plugins": [
        "espower",
        "transform-runtime",
        "transform-pipeline"
      ]
    }
```

in nuxt.config.js, .babelrc (without the "babel" key), and package.json. Yet I'm still getting "Syntax Error: Unexpected Token"

Any advise on how to get this working with Nuxt would be appreciated. I've tried restarting the server.

<div align="right"><sub><em>This question is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c1596">#c1596</a>)</em></sub></div>

**Comment by homerjam** [comment]
Have you tried `.babelrc`?

**Comment by babakness** [comment]
Hi James, yes I've tried all of `.babelrc`, `package.json`, and `nuxt.config.js`

**Comment by homerjam** [comment]
Hmmm. Does nuxt work without the pipeline operator in the code?

**Comment by homerjam** [comment]
How about `extend()`ing the webpack config?

```
build: {
    extend(config, ctx) {
      if (ctx.dev && ctx.isClient) {
        config.module.rules.push({
          enforce: 'pre',
          test: /\.(js|vue)$/,
          loader: 'eslint-loader',
          exclude: /(node_modules)/,
        });
      }
}
```

**Comment by calebanthony** [comment]
@homerjam is right about the `extend()`. Just adding in the `.babelrc` isn't enough, you also have to tell Webpack to use it.

So first make sure your `.babelrc` is set up. Then, in your `nuxt.config.js`:

```javascript
build: {
    extend (config, ctx) {
      if (ctx.isClient) {
        config.module.rules.push({
          enforce: 'pre',
          test: /\.(js|vue)$/,
          loader: 'eslint-loader',
          exclude: /(node_modules)/
        }, {
          test: /\.js$/,
          loader: 'babel-loader',
          exclude: /(node_modules)/
        })
      }
    }
}
```

**Comment by babakness** [comment]
Yeah, still doesn't work. Tried this:

```javascript
extend (config, ctx) {
        config.module.rules.push( {
          test: /\.js$/,
          loader: 'babel-loader',
          exclude: /(node_modules)/,
          options: {
            "presets": [
              "es2015",
              "stage-2"
            ],
            "plugins": [
              "espower",
              "transform-runtime",
              "transform-pipeline"
            ]
          }
        })
    }
```

No luck. The `|>` operator is still not being interpreted

**Comment by homerjam** [comment]
Have you managed to get this working outside of nuxt? Maybe you can pinpoint the problem more easily in isolation.

**Comment by christophwolff** [comment]
Any updates on this?

**Comment by benmccallum** [comment]
Nuxt's webpack build, including babel-loader is configured here:
https://github.com/nuxt/nuxt.js/blob/40aa0ce41e6bae56939a9559aedbaa1ff7c3e879/lib/webpack/base.config.js#L70

It seems to indicate that it doesn't use the .babelrc config file, as it sets option babelrc: false. Instead, it merges whatever is set in nuxt.config.js > babel with what they have inline there. Don't know whether this is new or not.

TL;DR; use nuxt.config.js > babel property. e.g. `build: { babel: { presets: ['es2015', 'stage'] } },`

**Comment by mr-e-** [comment]
I am having the same issue. I have included the babel spread operator in my nuxt.config:

`build: {
    babel: {
      'plugins': ['transform-object-rest-spread']
    }
}`

and I still get a compile error Unexpected Token. I too have tried using a .babelrc and I also get errors regarding the spread operator on internal nuxt files. If nuxt is using spread internally then why can't we?

**Comment by lenvonsam** [comment]
any solution for babel plugin in nuxt.config.js

**Comment by Merovex** [comment]
I see @clarkdo closed this by changing it to where .babelrc can now be used. However, is this actually fixed?

**Comment by usb248** [comment]
I have the same issue ... I don't understand why this issue is closed while nobody can use babel plugin in nuxt ...

**Comment by clarkdo** [comment]
Sorry for closing this issue, if you can still reproduce this issue, could you please provide a simple repo ?

**Comment by benmccallum** [comment]
Can you explain what you mean by using the custom server? Are you saying if we pick any of the other server options when we setup with the cli that babel won't execute as part of the build/startup build? 

Also, what's the benefits of using a custom server (like express) over the default? I'd definitely just use the default if I didn't need anything else and it was good enough.

**Comment by clarkdo** [comment]
I mean if you're using the custom server, the source codes belong to your server, such as `router`, `api`, `model` and so on, they won't be transpiled by nuxt, because these codes are outside the nuxt.

Hope you can understand my poo English, sorry for this.

**Comment by janzheng** [comment]
@clarkdo is right, this is how I got object-spread to work:

First I made sure to install the babel plugins I needed, by running:
`npm install --save-dev babel-cli babel-preset-es2015` and 
`npm install --save-dev babel-plugin-transform-object-rest-spread `

Then I added `babel` to the `build` section, specifying what presets and plugins to use, and then I added `babel-loader` like mentioned above:
```
build: {
    babel: {
      presets: ['es2015', 'stage-2'],
      plugins: ["transform-vue-jsx", "transform-runtime", "transform-object-rest-spread"],
    },
    extend (config, { isDev, isClient }) {
      if (isDev && isClient) {
        config.module.rules.push(
        {
          enforce: 'pre',
          test: /\.(js|vue)$/,
          loader: 'eslint-loader',
          exclude: /(node_modules)/
        },{
          test: /\.js$/,
          loader: 'babel-loader',
          exclude: /(node_modules)/
        })
      }
    }
}
```

Then I also had to add the `.babelrc` file with all the same plugins specified in there.

Hope this helps!

**Comment by benmccallum** [comment]
Definitely good to see someone with success. I haven't really revisited this. But from your example and clarkdo's comments, it seems like that handles transpiration of files in your own app. For me I had an issue getting transpiration occurring on bootstrap JS files in it's node_modules folder, likely because it was being excluded by the default nuxt rule. Im sure it was just a matter of tweaking the extend like you did to setup your own exclude/include and make sure they transpire, I just never had enough time to get it working with what I tried.

**Comment by devunion** [comment]
@clarkdo Could you attach your .babelrc file also? So, we will have complete solution in one place.

**Comment by clarkdo** [comment]
@devunion 
The babelrc depends on which feature you want to use in you server code, here is an example:

```json
{
  "presets": ["es2015"],
  "plugins": ["transform-runtime",
    "transform-decorators-legacy",
    "transform-class-properties",
    "transform-object-rest-spread",
    "transform-async-to-generator", ["component", [{
      "libraryName": "element-ui",
      "styleLibraryName": "theme-default"
    }]]
  ],
  "comments": false,
  "env": {
    "development": {
      "retainLines": true,
      "sourceMaps": true,
      "comments": true
    }
  }
}
```

**Comment by rnenjoy** [comment]
Same here, tries to use:

```
{
  "presets": [
    ["env", {
      "targets": {
        "node": "9.3.0"
      }
    }]
  ],
  "plugins": [
    "babel-plugin-inline-import",
    ["transform-imports", {
      "vuetify": {
        "transform": "vuetify/es5/components/${member}",
        "preventFullImport": true
      }
    }]
  ]
}
```

with nuxt-edge, but its not respected.

**Comment by clarkdo** [comment]
@babakness Since this issue doesn't have any new information for a long time, I'll close it.

Nuxt 2 is going to release and upgrade to babel 7, you can have a glance at nuxt-edge 😸

**Comment by Merovex** [comment]
This should not have been closed. The defect remains, and there is an example that appears testable.

**Comment by clarkdo** [comment]
@menthol Could you provide a repo to reproduce the issue ?

**Comment by menthol** [comment]
@clarkdo ???

**Comment by clarkdo** [comment]
Sorry @Merovex

**Comment by shirohana** [comment]
I'm launching a new project with Nuxt, I **can** use `babel-plugin-transform-flow-strip-types` in my case. Does it helps?

#### package.json
```js
  ...
  "dependencies": {
    "nuxt": "^1.4.2"
  },
  "devDependencies": {
    "babel-plugin-transform-flow-strip-types": "^6.22.0",
    "flow-bin": "^0.79.1"
  }
  ...
```

#### nuxt.config.js
```js
module.exports = {
  build: {
    babel: {
      plugins: ['transform-flow-strip-types']
    }
  }
}
```

above: w/o `plugins: ['transform-flow-strip-types']` in nuxt.config.js
below: w/ `plugins: ['transform-flow-strip-types']`

![2018-08-27 2 13 19](https://user-images.githubusercontent.com/11359892/44643892-b9e23180-aa04-11e8-8a71-90707d4d816e.png)

**Comment by clarkdo** [comment]
In Nuxt 1.0, if you want to use .babelrc, I think need to config `build.babel.babelrc: false`

Now `babel-loader` has changed the logic, we need to use extend for babelrc file
Some thing like : `build.babel.extends: path.join(__dirname, "fixtures/babelrc")`,

https://github.com/babel/babel-loader/blob/master/src/index.js#L70-L73

**Comment by JkmAS** [comment]
I have the same issue. I want to use `babel-plugin-root-import` for server side in nuxt-edge to support `~/path/file` imports. But it seems, that the `babelrc` is ignored.

**Comment by manniL** [comment]
An interactive CodeSandbox that shows how to use the [pipeline-proposal](https://www.npmjs.com/package/@babel/plugin-proposal-pipeline-operator) babel plugin with Nuxt 2 can be found [here](https://codesandbox.io/s/18j0jkn72q).

**Comment by Extarys** [comment]
@manniL the codesandbox returns `Something went wrong`
